File: src/main/java/arcana/client/events/RenderEvents.java
package arcana.client.events;

import arcana.common.aspects.AspectList;
import arcana.common.aspects.AspectStack;
import arcana.common.aspects.AspectUtils;
import arcana.common.aspects.ItemAspectRegistry;
import arcana.common.capability.Marks;
import arcana.common.particles.MarkParticle;
import arcana.utils.ClientUtil;
import com.mojang.blaze3d.systems.RenderSystem;
import net.minecraft.client.Minecraft;
import net.minecraft.client.entity.player.ClientPlayerEntity;
import net.minecraft.client.gui.screen.Screen;
import net.minecraft.client.world.ClientWorld;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.text.StringTextComponent;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.client.event.RenderTooltipEvent;
import net.minecraftforge.client.event.RenderWorldLastEvent;
import net.minecraftforge.event.entity.player.ItemTooltipEvent;
import net.minecraftforge.eventbus.api.EventPriority;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.common.Mod;

import javax.annotation.Nonnull;

@Mod.EventBusSubscriber(value = Dist.CLIENT)
public class RenderEvents {
    @SubscribeEvent
    static void marks(RenderWorldLastEvent event) {
        ClientWorld world = Minecraft.getInstance().level;
        ClientPlayerEntity player = Minecraft.getInstance().player;
        if (world == null || player == null)
            return;
        if (player.tickCount % 15 != 0)
            return;
        Marks cap = world.getCapability(Marks.CAPABILITY).resolve().orElse(null);
        for (int i = 0; i < AspectUtils.primalAspects.length; i++) {
            for (BlockPos pos : cap.positions[i]) {
                for (int j = 0; j < 3; j++)
                    world.addParticle(new MarkParticle.Data(AspectUtils.primalAspects[i]),
                        pos.getX() + world.random.nextFloat(), pos.getY() + 0.3, pos.getZ() + world.random.nextFloat(),
                        0, 0.1, 0);
            }
        }
    }

    @SubscribeEvent(priority = EventPriority.LOWEST)
    public static void onRenderTooltipPost(@Nonnull RenderTooltipEvent.PostText event) {
        if (!Screen.hasShiftDown())
            return;
        AspectList stacks = ItemAspectRegistry.get(event.getStack());
        if (stacks.list.size() > 0) {
            RenderSystem.pushMatrix();
            RenderSystem.translatef(0, 0, 500);
            RenderSystem.color3f(1F, 1F, 1F);
            Minecraft mc = Minecraft.getInstance();
            RenderSystem.translatef(0F, 0F, mc.getItemRenderer().blitOffset);

            int x = event.getX();
            int y = 10 * (event.getLines().size() - 3) + 14 + event.getY();
            for (AspectStack as : stacks) {
                ClientUtil.renderAspectStack(event.getMatrixStack(), as.getAspect(), as.amount, x, y);
                x += 20;
            }
            RenderSystem.popMatrix();
        }
    }

    @SubscribeEvent(priority = EventPriority.LOWEST)
    public static void makeTooltip(@Nonnull ItemTooltipEvent event) {
        if (!Screen.hasShiftDown())
            return;
        AspectList stacks = ItemAspectRegistry.get(event.getItemStack());
        if (stacks.list.size() > 0) {
            // amount of spaces that need inserting
            int filler = stacks.list.size() * 5;
            // repeat " " *filler
            StringBuilder sb = new StringBuilder();
            for (int __ = 0; __ < filler; __++) {
                String s = " ";
                sb.append(s);
            }
            String collect = sb.toString();
            event.getToolTip().add(new StringTextComponent(collect));
            event.getToolTip().add(new StringTextComponent(collect));
        }
    }
}

File: src/main/kotlin/arcana/client/events/RenderEvents.kt
package arcana.client.events


import arcana.common.aspects.AspectList
import arcana.common.aspects.Aspects
import arcana.common.aspects.ItemAspectRegistry
import arcana.common.capability.getMarks
import arcana.common.particles.MarkParticle
import arcana.utils.ClientUtil
import com.mojang.blaze3d.systems.RenderSystem
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.screen.Screen
import net.minecraft.util.text.StringTextComponent
import net.minecraftforge.api.distmarker.Dist
import net.minecraftforge.client.event.RenderTooltipEvent.PostText
import net.minecraftforge.client.event.RenderWorldLastEvent
import net.minecraftforge.event.entity.player.ItemTooltipEvent
import net.minecraftforge.eventbus.api.EventPriority
import net.minecraftforge.eventbus.api.SubscribeEvent
import net.minecraftforge.fml.common.Mod.EventBusSubscriber
import javax.annotation.Nonnull

@EventBusSubscriber(value = [Dist.CLIENT])
object RenderEvents {
    @SubscribeEvent
    fun marks(event: RenderWorldLastEvent) {
        val world = Minecraft.getInstance().level
        val player = Minecraft.getInstance().player
        if (world == null || player == null)
            return
        if (player.tickCount % 15 != 0)
            return
        val marks = world.getMarks()
        for (i in Aspects.PRIMAL.indices) {
            for (pos in marks.positions[i]) {
                for (j in 0..2)
                    world.addParticle(
                        MarkParticle.Data(Aspects.PRIMAL[i]),
                        pos.x + world.random.nextDouble(), pos.y + 0.3, pos.z + world.random.nextDouble(),
                        0.0, 0.1, 0.0
                )
            }
        }
    }

    private var line = 0
    @SubscribeEvent(priority = EventPriority.LOWEST)
    fun onRenderTooltipPost(@Nonnull event: PostText) {
        if (!Screen.hasShiftDown()) return
        val stacks: AspectList = ItemAspectRegistry.get(event.stack)
        if (stacks.list.size > 0) {
            RenderSystem.pushMatrix()
            RenderSystem.translatef(0f, 0f, 500f)
            RenderSystem.color3f(1f, 1f, 1f)
            val mc = Minecraft.getInstance()
            RenderSystem.translatef(0f, 0f, mc.itemRenderer.blitOffset)
            var x = event.x
            val y = 10 * line + 2 + event.y
            for (ass in stacks) {
                ClientUtil.renderAspectStack(event.matrixStack, ass.aspect, ass.amount, x, y)
                x += 20
            }
            RenderSystem.popMatrix()
        }
    }

    @SubscribeEvent(priority = EventPriority.LOWEST)
    fun makeTooltip(@Nonnull event: ItemTooltipEvent) {
        if (!Screen.hasShiftDown()) return
        val stacks: AspectList = ItemAspectRegistry.get(event.itemStack)
        if (stacks.list.size > 0) {
            line = event.toolTip.size
            // amount of spaces that need inserting
            val filler = stacks.list.size * 5
            // repeat " " *filler
            val sb = StringBuilder()
            for (i in 0 until filler) {
                val s = " "
                sb.append(s)
            }
            val collect = sb.toString()
            event.toolTip.add(StringTextComponent(collect))
            event.toolTip.add(StringTextComponent(collect))
        }
    }
}